v2/internal/testcommon/vcr/v3/replay_roundtripper_test.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package v3
+
+import (
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"testing"
+
+	. "github.com/onsi/gomega"
+
+	"github.com/go-logr/logr"
+
+	"github.com/Azure/azure-service-operator/v2/internal/testcommon/vcr"
+)
+
+func TestReplayRoundTripperRoundTrip_GivenSingleGET_ReturnsMultipleTimes(t *testing.T) {
+	t.Parallel()
+	g := NewGomegaWithT(t)
+
+	// Arrange
+	req := &http.Request{
+		URL:    &url.URL{Path: "/foo"},
+		Method: http.MethodGet,
+		Body:   io.NopCloser(strings.NewReader("GET body goes here")),
+	}
+
+	resp := &http.Response{
+		StatusCode: 200,
+	}
+
+	fake := vcr.NewFakeRoundTripper()
+	fake.AddResponse(req, resp)
+
+	// Act
+	replayer := NewReplayRoundTripper(fake, logr.Discard())
+
+	// Assert - first request works
+	//nolint:bodyclose // there's no actual body in this response to close
+	resp, err := replayer.RoundTrip(req)
+	g.Expect(err).ToNot(HaveOccurred())
+	g.Expect(resp.StatusCode).To(Equal(200))
+
+	// Assert - second request works by replaying the first
+	//nolint:bodyclose // there's no actual body in this response to close
+	resp, err = replayer.RoundTrip(req)
+	g.Expect(err).ToNot(HaveOccurred())
+	g.Expect(resp.StatusCode).To(Equal(200))
+
+	// Assert - third request works by replaying the first
+	//nolint:bodyclose // there's no actual body in this response to close
+	resp, err = replayer.RoundTrip(req)
+	g.Expect(err).ToNot(HaveOccurred())
+	g.Expect(resp.StatusCode).To(Equal(200))
+}
+
+func TestReplayRoundTripperRoundTrip_GivenSinglePUT_ReturnsOnceExtra(t *testing.T) {
```

Comment:
Should we have a test that ensures that the hash-based lookup works?

Specifically, something where there are 2 puts w/ different hashes and the first one fails, assert we get the first ones response replayed (not the second), and then another PUT which should give the second response?

Reply:
Though actually now that I re-read how the PUT logic you have works, I am not actually sure such a test will pass because the hash isn't tightly integrated into the `replayer.inner.RoundTrip` - I think it needs to be for the above to work?

Happy to make this comment nonblocking though and we can come around again and fix that issue? wdyt?

Reply:
I added another test, and I think the functionality is working as you desire. Let me know and I can address in a future PR if not.

Reply:
What I was thinking we wanted was this:

- PUT1 (served from recording)
- PUT2 (served from hash-based cache)
- PUT3 (served from recording)
- PUT4 (served from recording)

and what we have is

- PUT1 (served from recording)
- PUT2 (served from hash-based cache)
- PUT3 (served from hash-based cache)
- PUT4 (served from hash-based cache)

The difference being that once the first miss happens we're in the "served from hash-based cache" for the rest of that PUT. BUT, you're right that what we're doing actually meshes with what you said we'd do which is allow 1 miss per PUT-URL (not per PUT-URL-Hash).

I've approved this PR - I had missed the fact that even though we're playing the other hashes from the cache after the first miss it will work. If you feel that this option is valuable to note as a future change if needed, feel free to leave a TODO or non-TODO comment in the roundtripper that alludes to this as an option if we end up needing it. I suspect we will not end up needing it as we don't see that many flakes now so even the limit of 1 per URL is likely gonna be OK.

Reply:
> BUT, you're right that what we're doing actually meshes with what you said we'd do which is allow 1 miss per PUT-URL (not per PUT-URL-Hash).

What we're doing is allowing one miss per PUT body hash, not per URL (that changed when we switched to using body-hash as the key).

So the scenario you describe will work:

* PUT 1 - body1 (response served from recording)
* PUT 2 - body1 (response served from hash-based cache)
* PUT 3 - body2 (served from recording)
* PUT 4 - body2 (served from recording)